keystore/v2/keystore/filesystem/backend/file_lock.go

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020, Cossack Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package backend
+
+import (
+	"os"
+	"sync"
+	"syscall"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// fileLock is an interprocess read-write lock that serializes access to filesystem.
+//
+// This lock is implemented with BSD flock(2) and has corresponding semantics.
+//
+// Note that this is an advisory lock. That is, this lock can be used to guard
+// access to some filesystem resource by cooperating processes, but it does not
+// guarantee correct serialization of filesystem accesses by itself.
+//
+// Also note that flock(2) has subtly different semantics from POSIX record locks
+// created by fcntl(2). Read corresponding system manual pages for details.
+type fileLock struct {
```

Comment:
you know, this fileLock looks like well done code which good to publish at least in our code (make it public) to re-use, at most make this as a separate repository/module in github to allow use it other people too )

Reply:
Yeah, it's probably a good idea. I did not feel that it's generic enough, as it's mostly based around our use case for key stores which results in effectively one file lock per process.

I guess there are libraries that provide file locking. I shopped around right now and [this one](https://pkg.go.dev/github.com/gofrs/flock?tab=doc) comes up, for example. Initially i thought that we can get away with something “simple” for our use case, but now that I look at all these things I think that maybe we should just add a dependency since that code is likely to be way more tested than this hack.

Reply:
Oh, I've read the source code of that library more closely and it seems it does not protect from the case that caused this PR. It still allows the flock to be accidentally upgraded or downgraded and relies entirely on the system call to provide synchronization. No luck!

Reply:
As for publishing this, I'll keep that in mind but I think it's too early.

I'd wait for a second legitimate use case in Acra code base before even exporting it. Maybe that use case will get us to improve the lock in some way. Having it exported from `keystore/v2/filesystem/backend` does not really increase discoverability, and I feel that this lock is too task-specific to be placed into `utils`.

As for making it into a library, I'm flattered and I think it's a good idea to open-source reusable bits. However, we'd probably need to look around more to see if there are *really* no similar libraries*, prepare the repository, write a lot of tests, etc. This will take time so it's obviously out of scope for this PR. I've created a task to track that though.

<sub>* So far I've found [1](https://github.com/gofrs/flock), [2](https://github.com/MichaelS11/go-file-lock), [3](https://github.com/golang/go/tree/master/src/cmd/go/internal/lockedfile/internal/filelock), all different and none of them is combined with in-process synchronization.</sub>